Add GitHub build action with app compilation

Would be cool to have apk generated on each commit.

It seems you can check
https://github.com/sds100/KeyMapper/blob/master/.github/workflows/android.yml as example.

P.S. I made a quick try https://github.com/bahusoid/orion-viewer but build quite expectedly [fails](https://github.com/bahusoid/orion-viewer/runs/649297206) as I have zero knowledge in android development.

> Thank you, I will look, but as I understand there would be unsigned build and it would conflict with installed signed application

Not sure. In case of  Key Mapper it's compiled under different name so debug apk is installed as different app.


> you need 2 task to be executed

Thanks. But really I'm not the right person to implement it properly. I tried to modify build script with you suggestion. But it seems both compilation steps have some issues (see full [build log](https://github.com/bahusoid/orion-viewer/runs/650051965)):
1) `Prepare thirdparty` step  seems to be successful but last text in log is very suspicious:
```
thirdparty/freeglut/include/GL/freeglut_std.h:143:13: fatal error: GL/gl.h: No such file or directory
 #   include <GL/gl.h>
             ^~~~~~~~~
compilation terminated.
make: *** [build/release/platform/gl/gl-annotate.o] Error 1

BUILD SUCCESSFUL in 3m 30s
4 actionable tasks: 4 executed
```
2) `Build debug apk` also fails. And if I got it right due to the following issue:
```
A problem occurred configuring project ':nativeLibs:djvuModule'.
NdkCompile is no longer supported
> No version of NDK matched the requested version 20.1.5948944. Versions available locally: 21.0.6113669
```
Have no idea what make of it and how to fix it...

Still no luck: https://github.com/bahusoid/orion-viewer/runs/650324027?check_suite_focus=true

Now error somewhere inside mupdf:
```
2020-05-06T17:17:24.7838063Z > Task :nativeLibs:mupdfModule:externalNativeBuildDebug FAILED
2020-05-06T17:17:24.7838336Z [348/349] Linking C static library libmupdf_core.a
2020-05-06T17:17:24.7839380Z [349/349] Linking CXX shared library /home/runner/work/orion-viewer/orion-viewer/nativeLibs/mupdfModule/build/intermediates/cmake/debug/obj/armeabi-v7a/libmupdf_java.so
2020-05-06T17:17:24.7839995Z FAILED: /home/runner/work/orion-viewer/orion-viewer/nativeLibs/mupdfModule/build/intermediates/cmake/debug/obj/armeabi-v7a/libmupdf_java.so 
2020-05-06T17:17:24.7841953Z : && /usr/local/lib/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=armv7-none-linux-androideabi16 --gcc-toolchain=/usr/local/lib/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/usr/local/lib/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security   -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc_real.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--fatal-warnings -Wl,--exclude-libs,libunwind.a -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libmupdf_java.so -o /home/runner/work/orion-viewer/orion-viewer/nativeLibs/mupdfModule/build/intermediates/cmake/debug/obj/armeabi-v7a/libmupdf_java.so CMakeFiles/mupdf_java.dir/home/runner/work/orion-viewer/orion-viewer/nativeLibs/common/orion_bitmap.c.o CMakeFiles/mupdf_java.dir/home/runner/work/orion-viewer/orion-viewer/nativeLibs/common/list.c.o CMakeFiles/mupdf_java.dir/mupdf_native.c.o  libmupdf_core.a libmupdf_thirdparty.a -ljnigraphics -llog -lm -Wl,--gc-sections -latomic -lm && :
2020-05-06T17:17:24.7842799Z /home/runner/work/orion-viewer/orion-viewer/nativeLibs/mupdf/source/fitz/noto.c:0: error: undefined reference to '_binary_NimbusMonoPS_Regular_cff_size'
2020-05-06T17:17:24.78433
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -1,0 +1,29 @@
+name: Android CI
+
+on: [push, pull_request]
+
+jobs:
+  apk:
+    name: Generate APK
+    runs-on: ubuntu-18.04
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Create empty local.properties
+        uses: finnp/create-file-action@1.0.0
+        env:
+          FILE_NAME: "local.properties"
+          FILE_DATA: ""
+      - name: Prepare thirdparty
+        run: bash ./gradlew -b thirdparty_build.gradle downloadAndPatchDjvu downloadAndMakeMupdf
+      - name: Build debug APK
+        run: bash ./gradlew assembleCi --stacktrace -Porion.CIBuild=true
+      - name: Upload APK
+        uses: actions/upload-artifact@v2
+        with:
+          name: OrionViewer
+          path: orion-viewer/build/outputs/apk/**/*.apk

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -20,5 +20,5 @@ jobs:
       - name: Upload APK
         uses: actions/upload-artifact@v2
         with:
-          name: OrionViewer
+          name: OrionViewer-${{github.run_number}}
           path: orion-viewer/build/outputs/apk/**/*.apk

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -13,11 +13,6 @@ jobs:
         uses: actions/setup-java@v1
         with:
           java-version: 1.8
-      - name: Create empty local.properties
-        uses: finnp/create-file-action@1.0.0
-        env:
-          FILE_NAME: "local.properties"
-          FILE_DATA: ""
       - name: Prepare thirdparty
         run: bash ./gradlew -b thirdparty_build.gradle downloadAndPatchDjvu downloadAndMakeMupdf
       - name: Build debug APK

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -20,5 +20,5 @@ jobs:
       - name: Upload APK
         uses: actions/upload-artifact@v2
         with:
-          name: OrionViewer-${{github.run_number}}
+          name: OrionViewer-CI-build${{github.run_number}}
           path: orion-viewer/build/outputs/apk/**/*.apk

--- a/build.gradle
+++ b/build.gradle
@@ -29,7 +29,7 @@ project.ext {
     orionCompileSdk = 28
     orionBuildToolsVersion = "28.0.3"
     if (isCIBuild) {
-        orionNdkVersion = "21.0.6113669"
+        orionNdkVersion = "21.1.6352462"
     }
     else {
         orionNdkVersion = "20.1.5948944"

--- a/thirdparty_build.gradle
+++ b/thirdparty_build.gradle
@@ -76,7 +76,7 @@ task downloadAndMakeMupdf(dependsOn: downloadMupdf)  {
         /*For more details see mupdf/platform/android/viewer/ReadMe.txt*/
         exec {
             workingDir file(mupdf)
-            commandLine 'make generate'
+            commandLine 'make', 'generate'
         }
     }
 }

